core/devices/machine_topology.cpp

```diff
@@ -0,0 +1,349 @@
+    hwloc_obj_type_t type,
+    std::tuple<std::vector<MachineTopology::normal_obj_info>,
+               MachineTopology::hwloc_manager<hwloc_bitmap_s>> &objects)
+{
+#if GKO_HAVE_HWLOC
+    std::get<1>(objects) =
+        hwloc_manager<hwloc_bitmap_s>(init_bitmap(), hwloc_bitmap_free);
+    // Get the number of normal objects of a certain type (Core, PU, Machine
+    // etc.).
+    unsigned num_objects = hwloc_get_nbobjs_by_type(this->topo_.get(), type);
+    for (unsigned i = 0; i < num_objects; i++) {
+        // Get the actual normal object of the given type.
+        hwloc_obj_t obj = hwloc_get_obj_by_type(this->topo_.get(), type, i);
+        std::get<0>(objects).push_back(
+            normal_obj_info{obj, obj->logical_index, obj->os_index,
+                            obj->gp_index, hwloc_bitmap_first(obj->nodeset)});
+    }
+#endif
+}
+
+
+inline int MachineTopology::get_obj_local_id_by_os_index(
+    std::vector<MachineTopology::normal_obj_info> &objects,
+    size_type os_index) const
+{
+#if GKO_HAVE_HWLOC
+    for (auto id = 0; id < objects.size(); ++id) {
+        if (objects[id].os_id == os_index) {
+            return id;
+        }
+    }
+    return -1;
+#endif
+}
+
+
+inline int MachineTopology::get_obj_local_id_by_gp_index(
+    std::vector<MachineTopology::normal_obj_info> &objects,
+    size_type gp_index) const
+{
+#if GKO_HAVE_HWLOC
+    for (auto id = 0; id < objects.size(); ++id) {
+        if (objects[id].gp_id == gp_index) {
+            return id;
+        }
+    }
+    return -1;
+#endif
+}
+
+
+void MachineTopology::load_objects(
+    hwloc_obj_type_t type, std::vector<MachineTopology::io_obj_info> &vector)
+{
+#if GKO_HAVE_HWLOC
+    GKO_ASSERT(std::get<0>(this->cores_).size() != 0);
+    GKO_ASSERT(std::get<0>(this->pus_).size() != 0);
+    unsigned num_objects = hwloc_get_nbobjs_by_type(this->topo_.get(), type);
+    for (unsigned i = 0; i < num_objects; i++) {
+        // Get the actual PCI object.
+        hwloc_obj_t obj = hwloc_get_obj_by_type(this->topo_.get(), type, i);
+        // Get the non-IO ancestor (which is the closest and the one that can be
+        // bound to) of the object.
+        auto ancestor = hwloc_get_non_io_ancestor_obj(this->topo_.get(), obj);
+        // Create the object.
+        vector.push_back(
+            io_obj_info{obj, obj->logical_index, obj->os_index, obj->gp_index,
+                        hwloc_bitmap_first(ancestor->nodeset), ancestor});
+        // Get the corresponding cpuset of the ancestor nodeset
+        hwloc_cpuset_t ancestor_cpuset = hwloc_bitmap_alloc();
+        hwloc_cpuset_from_nodeset(this->topo_.get(), ancestor_cpuset,
+                                  ancestor->nodeset);
+        // Find the cpu object closest to this device from the ancestor cpuset
+        // and store its id for binding purposes
+        vector.back().closest_cpu_id = get_obj_local_id_by_os_index(
+            std::get<0>(this->pus_), hwloc_bitmap_first(ancestor_cpuset));
```

Comment:
Here I think it would be better to simply store all the IDs as explained in other places.